### ISSUE TITLE
Climb evaluation: ObjectSet_test_0 consistently fails (~2%) while test_1/test_2 succeed (~100%)

### ISSUE THREAD

**Issue by Luoyuchen0704** [liangpan99/TokenHSI#17]
# Climb evaluation: ObjectSet_test_0 consistently fails (~2%) while test_1/test_2 succeed (~100%)

## Environment
- TokenHSI framework, multi-task setting (HumanoidTrajSitCarryClimb)
- Checkpoint: `ckpt_stage1.pth` (the official pretrained multi-task checkpoint)
- Config: `amp_humanoid_traj_sit_carry_climb.yaml` (default multi-task)
- Eval command: `python tokenhsi/run.py --task HumanoidTrajSitCarryClimb --cfg_train ... --cfg_env ... --checkpoint output/tokenhsi/ckpt_stage1.pth --test --num_envs 512 --headless --eval --eval_task climb --seed 0`
- Seed 0, StateInit = Random, all 3 object categories (Box, Cabinet, Table_Square)

## Observed Behavior

When evaluating the climb task with `--seed 0`, the 3 eval repeats show a stark and **deterministically reproducible** performance gap:

| Repeat | Seed | Success Rate | Terminations/512 | Avg Precision |
|--------|------|-------------|-----------------|---------------|
| test_0 | 0+0=0 | **~2%** | **~286 (56%)** | 5.3cm |
| test_1 | 0+1=1 | **~100%** | 0-3 | 4.1cm |
| test_2 | 0+2=2 | **~100%** | 0-2 | 4.1cm |

Overall average: **~67%** (far from the reported 99.8% in the paper).

This pattern persists across:
- 8 independent eval runs over different days (May 18-19, June 17) — **always the same**
- `StateInit.Start` vs `StateInit.Random` — **no difference**
- Removing Cabinet objects from the test set (Box + Table only) — **no difference**
- Different base seeds (0, 5, 7) — always **one of the 3 repeats fails at ~2%** while the other two succeed at ~100%

## Paper vs Code Discrepancy

The paper states (Section 4.1):
> "The object is spawned between **1 m and 5 m** from the character with randomized orientation."

However, the actual implementation in `humanoid_climb.py:_reset_objects` (line 697) uses:
```python
root_pos_xy *= torch.rand(len(ids), 1, device=self.device) * 9.0 + 1.0  # 1~10m
```

This is **1-10m**, not 1-5m. But even after changing to 1-5m, the evaluation results remained the same (test_0 still ~2%), so distance is not the root cause.

## Paper's Evaluation Protocol

The paper reports (Section 4.1):
> "512 trials are collected to calculate the success rate."
> "Values are reported in the format of mean±std."
> "Colored regions denote mean values ± a standard deviation based on 3 models initialized with different random seeds."

This suggests the paper averages results across 3 seeds at the *training* level (3 model initializations), not at the eval level. The eval's 3-fold repeat with `base_seed + i` is a different mechanism.

## Investigation Summary

We ruled out these possible causes:

| Hypothesis | Test | Result |
|-----------|------|--------|
| Cabinet objects are too hard | Removed Cabinet from test set | Still 67% |
| Random initial motion frames cause unstable starts | Changed to StateInit.Start (time=0) | Still 67% |
| Distance distribution differs between repeats | Sampled distances for seeds 0/1/2 | Similar distributions |
| Random seed fluke | Ran with seeds 0, 5, 7 | Always 1 bad repeat + 2 good repeats |
| Checkpoint issue | Used the original `ckpt_stage1.pth` | Same across all 8 runs |

## Key Observation

Within a single eval run, `ObjectSet_test_0` consistently has ~286/512 terminations (the character falls before reaching the climbing position), while `test_1` and `test_2` have near-zero terminations. Since the only difference between repeats is the torch seed (`base_seed + i`), **repeat 0 somehow samples a set of initial conditions that leads to systematic failure**.

This affects the overall climb evaluation: single-seed evaluation gives ~67%, but if we only look at repeats 1 and 2 (seeds 1 and 2), the result would be ~100%.

## Questions

1. Is this a known issue? How was the 99.8% reported in the paper obtained — was a specific seed used or was the checkpoint trained differently?
2. The 3 eval repeats use `torch.manual_seed(base_seed + i)`, which controls motion sampling, object placement, and distance randomization. Are the reported results averaged over multiple seeds, or is a specific "good" seed used?
3. Is the `ckpt_stage1.pth` checkpoint the final converged model used for the paper's climb evaluation, or was a later/better checkpoint used?
4. Could the high termination rate in repeat 0 be related to how `torch.multinomial` or `torch.rand` interacts with the environment initialization under specific seed states?

## Reproducibility

```bash
python tokenhsi/run.py --task HumanoidTrajSitCarryClimb \
    --cfg_train tokenhsi/data/cfg/train/rlg/amp_imitation_task_transformer_multi_task.yaml \
    --cfg_env tokenhsi/data/cfg/multi_task/amp_humanoid_traj_sit_carry_climb.yaml \
    --motion_file tokenhsi/data/dataset_loco_sit_carry_climb.yaml \
    --checkpoint output/tokenhsi/ckpt_stage1.pth \
    --test --num_envs 512 --headless --eval --eval_task climb --seed 0
```

Expected output:
```
ObjectSet_test_0: success_rate ≈ 1-4%
ObjectSet_test_1: success_rate ≈ 97-100%
ObjectSet_test_2: success_rate ≈ 99-100%
Overall: ~67%
```

**Comment by liangpan99** [comment]
Thanks for the detailed investigation.

1. The 99.8% reported in the paper was computed from 3 independently trained Stage-1 checkpoints. Each checkpoint was trained with a different random seed, then evaluated with our fixed evaluation protocol. The final number in the paper is reported as mean ± std over these 3 trained models.

2. For each single checkpoint, the evaluation script runs 3 deterministic evaluation repeats: ObjectSet_test_0 / test_1 / test_2. During our own evaluation, we also observed that the first repeat can sometimes produce abnormal results, e.g., a near-0% success rate. Therefore, we designed the evaluation to run three repeats and used the stable ObjectSet_test_1 as the final evaluation result for that checkpoint. Averaging ObjectSet_test_0/1/2 is not how the paper’s Table 1 result was computed.

3. The released ckpt_stage1.pth is a pretrained Stage-1 multi-task checkpoint. However, the paper result was computed from 3 independently trained checkpoints, while currently only one checkpoint is released. Therefore, the released checkpoint alone cannot exactly reproduce the mean ± std number reported in the paper.

**Comment by Luoyuchen0704** [comment]
Thank you very much for the detailed and helpful response! This completely clarifies the evaluation protocol for us.
